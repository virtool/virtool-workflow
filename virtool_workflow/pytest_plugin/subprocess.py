import pytest

import virtool_workflow.runtime.run_subprocess


@pytest.fixture
def run_subprocess() -> virtool_workflow.runtime.run_subprocess.RunSubprocess:
    return virtool_workflow.runtime.run_subprocess.run_subprocess()
