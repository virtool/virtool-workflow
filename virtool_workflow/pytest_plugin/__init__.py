from virtool_workflow.pytest_plugin.data import data, Data
from virtool_workflow.pytest_plugin.subprocess import run_subprocess

__all__ = ["data", "Data", "run_subprocess"]
