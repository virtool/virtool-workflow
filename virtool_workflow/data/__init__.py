from virtool_workflow.data.analyses import analysis
from virtool_workflow.analysis.fastqc import fastqc
from virtool_workflow.data.hmms import hmms
from virtool_workflow.data.indexes import index
from virtool_workflow.data.jobs import job, push_status
from virtool_workflow.data.ml import ml
from virtool_workflow.data.samples import sample
from virtool_workflow.data.subtractions import subtractions
from virtool_workflow.data.uploads import uploads

__all__ = [
    "analysis",
    "fastqc",
    "hmms",
    "index",
    "job",
    "ml",
    "push_status",
    "sample",
    "subtractions",
    "uploads",
]
