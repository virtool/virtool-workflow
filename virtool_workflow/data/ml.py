"""A fixture and dataclass for working with machine learning models in workflows."""
from dataclasses import dataclass
from pathlib import Path

from pyfixtures import fixture
from structlog import get_logger
from virtool_core.models.ml import MLModelRelease

from virtool_workflow.api.client import APIClient
from virtool_workflow.data.analyses import WFAnalysis
from virtool_workflow.utils import make_directory

logger = get_logger("api")


@dataclass
class WFMLModelRelease:
    """
    A machine learning model.

    This class represents a machine learning model and the selected release of that
    model in the workflow.
    """

    id: int
    """The unique ID for the model release."""

    name: str
    """The name of the model release."""

    path: Path
    """The path to the model directory."""

    @property
    def file_path(self) -> Path:
        """The path to the model data file."""
        return self.path / "model.tar.gz"


@fixture
async def ml(
    _api: APIClient, analysis: WFAnalysis, work_path: Path
) -> WFMLModelRelease | None:
    if analysis.ml is None:
        return None

    model_id = analysis.ml.model.id
    model_release_id = analysis.ml.id

    log = logger.bind(model_id=analysis.ml.id, model_release_id=model_release_id)

    model_release_json = await _api.get_json(
        f"/ml/{model_id}/releases/{model_release_id}"
    )
    model_release = MLModelRelease(**model_release_json)

    log.info("Fetched ML model release json")

    release = WFMLModelRelease(
        id=model_release.id,
        name=model_release.name,
        path=work_path / "ml" / str(model_release.model.id) / str(model_release_id),
    )

    await make_directory(release.path)

    await _api.get_file(
        f"/ml/{model_id}/releases/{model_release_id}/model.tar.gz", release.file_path
    )

    log.info("Downloaded ML model data file")

    return release
