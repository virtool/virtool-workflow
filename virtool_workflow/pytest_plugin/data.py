import datetime
from dataclasses import dataclass
from pathlib import Path

import pytest
from pydantic_factories import ModelFactory, Use
from virtool_core.models.analysis import Analysis
from virtool_core.models.index import Index, IndexNested
from virtool_core.models.job import JobAcquired, JobMinimal, JobPing
from virtool_core.models.ml import MLModelRelease
from virtool_core.models.reference import Reference
from virtool_core.models.samples import (
    Sample,
    SampleNested,
)
from virtool_core.models.subtraction import (
    Subtraction,
    SubtractionNested,
    SubtractionFile,
)

from virtool_workflow.pytest_plugin.utils import SUBTRACTION_FILENAMES


@dataclass
class Data:
    analysis: Analysis
    """An analysis being populated in the active workflow."""

    index: Index
    """A finalized index to be used for testing analyses."""

    new_index: Index
    """An un-finalized index for testing index creation workflows."""

    job: JobAcquired
    """A fake job."""

    ml: MLModelRelease | None
    """An ML model release used in the active analysis."""

    reference: Reference
    """A reference to be used for testing analyses and index creation workflows."""

    sample: Sample
    """A finalized sample to be used for testing analyses."""

    new_sample: Sample
    """An un-finalized sample for testing sample creation workflows."""

    subtraction: Subtraction
    """A finalized subtraction to be used for testing analyses."""

    new_subtraction: Subtraction
    """An un-finalized subtraction for testing subtraction creation workflows."""


@pytest.fixture
def data(example_path: Path, static_datetime: datetime.datetime) -> Data:
    class AnalysisFactory(ModelFactory):
        __model__ = Analysis

        created_at = Use(lambda: static_datetime)
        updated_at = Use(lambda: static_datetime)

    class IndexFactory(ModelFactory[Index]):
        __model__ = Index

        created_at = Use(lambda: static_datetime)

    IndexFactory.seed_random(12)

    class JobFactory(ModelFactory):
        __model__ = JobAcquired

        created_at = Use(lambda: static_datetime)
        timestamp = Use(lambda: static_datetime)

    JobFactory.seed_random(55)

    class MLFactory(ModelFactory):
        __model__ = MLModelRelease

        created_at = Use(lambda: static_datetime)
        published_at = Use(lambda: static_datetime)

    class ReferenceFactory(ModelFactory):
        __model__ = Reference

        created_at = Use(lambda: static_datetime)

    ReferenceFactory.seed_random(22)

    class SampleFactory(ModelFactory):
        __model__ = Sample

        created_at = Use(lambda: static_datetime)
        removed_at = None
        uploaded_at = Use(lambda: static_datetime)

    SampleFactory.seed_random(5)

    class SubtractionFactory(ModelFactory):
        __model__ = Subtraction

        created_at = Use(lambda: static_datetime)

    job: JobAcquired = JobFactory.build()

    job.args = {
        "item_id": 1211,
        "resource_id": "foo",
        "test": True,
    }
    job.ping = JobPing(pinged_at=static_datetime)
    job.rights = {}

    """A finalized sample to be used for testing analyses."""
    sample = SampleFactory.build()
    sample.job = JobMinimal.parse_obj(job)
    sample.artifacts = []

    # A new sample with the fake job configured as the creation job for the sample.
    new_sample = SampleFactory.build()

    new_sample_job = JobFactory.build()
    new_sample_job.args["files"] = [
        {
            "id": 1,
            "name": "reads_1.fq.gz",
            "size": 100,
        },
        {
            "id": 2,
            "name": "reads_2.fq.gz",
            "size": 100,
        },
    ]
    new_sample_job.args["sample_id"] = new_sample.id

    new_sample.artifacts = []
    new_sample.job = JobMinimal.parse_obj(new_sample_job)
    new_sample.quality = None
    new_sample.reads = []
    new_sample.ready = False

    reference = ReferenceFactory.build()
    reference.targets = []

    index = IndexFactory.build()
    index.reference = Reference.parse_obj(reference)
    index.ready = True

    new_index: Index = IndexFactory.build()
    new_index.reference = Reference.parse_obj(reference)
    new_index.files = []
    new_index.ready = False

    ml: MLModelRelease = MLFactory.build()
    ml.ready = True
    ml.model.id = 5
    ml.id = 231

    subtraction = SubtractionFactory.build()
    subtraction.files = [
        SubtractionFile(
            download_url=f"/subtractions/{subtraction.id}/files/{filename}",
            id=(i + 1),
            name=filename,
            size=(example_path / "subtraction" / filename).stat().st_size,
            subtraction=subtraction.id,
            type="bowtie2",
        )
        for i, filename in enumerate(SUBTRACTION_FILENAMES)
    ]

    new_subtraction = SubtractionFactory.build()
    new_subtraction.files = []
    new_subtraction.ready = False

    analysis: Analysis = AnalysisFactory.build()
    analysis.sample = SampleNested.parse_obj(sample)
    analysis.workflow = "pathoscope_bowtie"
    analysis.index = IndexNested.parse_obj(index)
    analysis.reference = Reference.parse_obj(reference)
    analysis.subtractions = [SubtractionNested.parse_obj(subtraction)]

    return Data(
        analysis=analysis,
        index=index,
        new_index=new_index,
        job=job,
        ml=ml,
        reference=reference,
        sample=sample,
        new_sample=new_sample,
        subtraction=subtraction,
        new_subtraction=new_subtraction,
    )
