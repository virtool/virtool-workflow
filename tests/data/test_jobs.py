from pyfixtures import FixtureScope
from syrupy import SnapshotAssertion
from virtool_core.models.job import Job

from virtool_workflow.pytest_plugin.data import Data


async def test_ok(data: Data, scope: FixtureScope, snapshot: SnapshotAssertion):
    data.job.acquired = False

    job: Job = await scope.get_or_instantiate("job")

    assert job.dict() == snapshot(name="pydantic")
