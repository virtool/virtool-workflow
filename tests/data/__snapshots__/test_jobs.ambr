# name: test_ok[pydantic]
  dict({
    'acquired': False,
    'archived': False,
    'args': dict({
      'item_id': 1211,
      'resource_id': 'foo',
      'test': True,
    }),
    'created_at': datetime.datetime(2020, 1, 1, 1, 1, 1),
    'id': 'GMPXMVbyXHUfymCDaloV',
    'ping': dict({
      'pinged_at': datetime.datetime(2020, 1, 1, 1, 1, 1),
    }),
    'progress': 6380,
    'rights': dict({
    }),
    'stage': 'iPhGgaTklZTpXCJkWJYU',
    'state': <JobState.RUNNING: 'running'>,
    'status': list([
      dict({
        'error': dict({
          'details': list([
            'VCqTDwYzyKtSiEMpnbXp',
          ]),
          'traceback': list([
            'WhbIXmchuyDmybzRLozZ',
          ]),
          'type': 'ASTUagoKFMbsJQgaCJfJ',
        }),
        'progress': 3132,
        'stage': 'aOwDKGzVuuSkpRdRKwoO',
        'state': <JobState.ERROR: 'error'>,
        'step_description': None,
        'step_name': None,
        'timestamp': datetime.datetime(2020, 1, 1, 1, 1, 1),
      }),
    ]),
    'user': dict({
      'handle': 'RaOtEacjXkNWWrsBOfMP',
      'id': 'psihdpFaJrqQyqzKcYbM',
    }),
    'workflow': 'SbXerFVrWCqqjehJZicZ',
  })
# ---
